# Groover Data Engineering Project

By:

# <font color = 'gold'>Oluwatimileyin Victor ADEDIGBA

## DEMANDS:


1.   Enrich Data Warehouse with updated spotify_id and popularity score of artists
2.   SQL query to fetch: user_id, spotify_id, genres by artist, total number of genres artist is listed in and total number of artists listed in each genre.


# Data Loading

Installing the Required Libraries

In [ ]:
!pip install fuzzywuzzy -U -q PyDrive spotipy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 252.0/252.0 kB 5.1 MB/s eta 0:00:00


Importing the installed libraries

In [ ]:
import pandas as pd
import requests
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials
from fuzzywuzzy import fuzz
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials


/usr/local/lib/python3.10/dist-packages/fuzzywuzzy/fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


Authenticating Pydrive client to fetch data from drive

In [ ]:
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

To fetch the datasets, we create a function for it

In [ ]:
def download_file_from_drive(file_id, file_name):
    """
    Downloads a file from Google Drive using the provided-
    file ID and saves it with the given file name.

    """
    downloaded = drive.CreateFile({'id': file_id})

    # Fetching and saving locally
    downloaded.GetContentFile(file_name)


Using the function to fetch the files

In [ ]:
# Creating Id variables
artist_data_id = '1DEGJPcOXOk4Sf1r6kpi27_OvFdiusCUI'
spotify_data_id = '1N-ljIMIwDaMxZPmoaNeEc82dBR-8nsMv'
tag_art_id = '117KOhjS3MVz1DcsLh4yOE36RUhAd-XXQ'
tag_genre_id = '1Un_pZO2ZxwDsu8m4HbfEKtR7kw8w4WCV'

# Calling the download function
download_file_from_drive(artist_data_id, 'artist_data.csv')
download_file_from_drive(spotify_data_id, 'spotify_data.csv')
download_file_from_drive(tag_art_id, 'tag_artist_data.csv')
download_file_from_drive(tag_genre_id, 'tag_genre_data.csv')


Reading the data as DataFrames

In [ ]:
artist_data_df = pd.read_csv('artist_data.csv')
spotify_data_df = pd.read_csv('spotify_data.csv')
tag_artist_data_df = pd.read_csv('tag_artist_data.csv')
tag_genre_data_df = pd.read_csv('tag_genre_data.csv')


Displaying basic dataset outlook

In [ ]:
print(artist_data_df.head(), artist_data_df.shape)
print(spotify_data_df.head(), spotify_data_df.shape)
print(tag_artist_data_df.head(), tag_artist_data_df.shape)
print(tag_genre_data_df.head(), tag_genre_data_df.shape)


   user_id     artist_name
0    24824     tatum quinn
1   303724  jason kerrison
2    28027            k.b.
3   451943   milo x kahefa
4   215894          cutter (10000, 2)
   user_id              spotify_id
0    24824  3s0DwmaExsRr8KGfE8RkhH
1   303724  7iLGqGUSoPQtj8OH61HFwZ
2    28027  3Y0w7w2dUKo3DqPh2BDIxr
3   451943  70o5L3YGkE7lxVWPro3fkV
4   215894  0ti5jOsC9Jk3TjG5odyaps (10000, 2)
   user_id  tag_id
0    24824     601
1   303724     601
2    28027     267
3   451943     432
4   215894     593 (27301, 2)
              genre  tag_id
0           hip_hop     145
1              jazz     267
2               pop     432
3  electronic_music     531
4             metal     593 (11, 2)


# DATA CLEANING

Showing missing and duplicated data information

In [ ]:
# Creating dictionaries for organization
inconsistent_data = {
    'DataFrame': ['artist_data_df', 'spotify_data_df',
                  'tag_artist_data_df', 'tag_genre_data_df'],

    'Missing Data': [artist_data_df.isnull().sum().sum(),
                     spotify_data_df.isnull().sum().sum(),
                     tag_artist_data_df.isnull().sum().sum(),
                     tag_genre_data_df.isnull().sum().sum()],

    'Duplicate Data': [artist_data_df.duplicated().sum(),
                       spotify_data_df.duplicated().sum(),
                       tag_artist_data_df.duplicated().sum(),
                       tag_genre_data_df.duplicated().sum()]
}

# Converting the dictionary to a DataFrame
inconsistent_data_df = pd.DataFrame(inconsistent_data)

# Display the table
inconsistent_data_df


DataFrame  Missing Data  Duplicate Data
0      artist_data_df             2               0
1     spotify_data_df             0               0
2  tag_artist_data_df             0               0
3   tag_genre_data_df             0               0

<font color = 'gold'>At this point, there are two ways to deal with these missing information:

1.   To remove the rows from a composite (fact) table
2.   To fill the information using spotify_id and API

<font color = 'gold'>However, I am removing the row.

<font color = 'gold'>This is because the artist_names associated with our spotify_id, as we will see, do not necessarily correspond with what is available on spotify database.

Therefore, because we are concerened with our customers, I chose work with what we know and not populate our artist_names with information from spotify.</font>



Joining the DataFrames to get a fact table

In [ ]:
# First joining artist_name with spotify_id, pri key = user_id
'''
SQL = SELECT *
FROM artist_data_df
INNER JOIN spotify_data_df
ON artist_data_df.user_id = spotify_data_df.user_id
'''
artist_name_spotify_id_df = artist_data_df.merge(spotify_data_df,
                                  on='user_id', how='inner')
artist_name_spotify_id_df

user_id     artist_name              spotify_id
0       24824     tatum quinn  3s0DwmaExsRr8KGfE8RkhH
1      303724  jason kerrison  7iLGqGUSoPQtj8OH61HFwZ
2       28027            k.b.  3Y0w7w2dUKo3DqPh2BDIxr
3      451943   milo x kahefa  70o5L3YGkE7lxVWPro3fkV
4      215894          cutter  0ti5jOsC9Jk3TjG5odyaps
...       ...             ...                     ...
9995    18629           naeko  0pyq140jTsiXoxlv1KKYzK
9996    24484  sinatra royale  6avD4d96Qd6SHW70V0fawF
9997    18645  the rosy cross  0o6kFVOKDFH3MIJUVXkdq1
9998    18611   didikongstyle  5jnSp9sNhdHJynWddUTdU8
9999    18618        katcross  5wKycvCDB0qE5vvw8CE13M

[10000 rows x 3 columns]

In [ ]:
# Joining artist tag with genre, pri key = tag_id
'''
SQL = SELECT *
FROM tag_artist_data_df
INNER JOIN tag_genre_data_df
ON tag_artist_data_df.tag_id = tag_genre_data_df.tag_id
'''
artist_tag_genre_df = tag_artist_data_df.merge(tag_genre_data_df,
                                      on='tag_id', how='inner')

artist_tag_genre_df

user_id  tag_id   genre
0        24824     601  reggae
1       303724     601  reggae
2        69766     601  reggae
3         3005     601  reggae
4          474     601  reggae
...        ...     ...     ...
27296    18642     900    rock
27297    18647     900    rock
27298    18648     900    rock
27299    18645     900    rock
27300    18611     900    rock

[27301 rows x 3 columns]

In [ ]:
# Grouping all genres with their artists_id
'''
# SQL = SELECT user_id, GROUP_CONCAT(genre) AS genre
# FROM artist_tag_genre_df
# GROUP BY user_id
'''
grouped_genre_df = artist_tag_genre_df.groupby(
    'user_id')['genre'].agg(list).reset_index()

grouped_genre_df

user_id                             genre
0           9  [reggae, jazz, electronic_music]
1          12                      [soul, rock]
2          14                 [pop, funk, trap]
3          16          [electronic_music, soul]
4          23               [trap, disco, rock]
...       ...                               ...
9995   493812                [pop, disco, rock]
9996   493815   [metal, funk, electronic_music]
9997   493819     [pop, electronic_music, rock]
9998   493822                     [disco, soul]
9999   493825             [reggae, jazz, metal]

[10000 rows x 2 columns]

In [ ]:
# Combining all extracted infomation into a pre-warehouse
'''
# SQL = SELECT *
# FROM artist_name_spotify_id_df
# INNER JOIN grouped_genre_df
# ON artist_name_spotify_id_df.user_id = grouped_genre_df.user_id
'''
pre_warehouse_df = pd.merge(artist_name_spotify_id_df,
                        grouped_genre_df, on='user_id', how='inner')

pre_warehouse_df


user_id     artist_name              spotify_id  \
0       24824     tatum quinn  3s0DwmaExsRr8KGfE8RkhH   
1      303724  jason kerrison  7iLGqGUSoPQtj8OH61HFwZ   
2       28027            k.b.  3Y0w7w2dUKo3DqPh2BDIxr   
3      451943   milo x kahefa  70o5L3YGkE7lxVWPro3fkV   
4      215894          cutter  0ti5jOsC9Jk3TjG5odyaps   
...       ...             ...                     ...   
9995    18629           naeko  0pyq140jTsiXoxlv1KKYzK   
9996    24484  sinatra royale  6avD4d96Qd6SHW70V0fawF   
9997    18645  the rosy cross  0o6kFVOKDFH3MIJUVXkdq1   
9998    18611   didikongstyle  5jnSp9sNhdHJynWddUTdU8   
9999    18618        katcross  5wKycvCDB0qE5vvw8CE13M   

                                  genre  
0                  [reggae, funk, trap]  
1                  [reggae, jazz, trap]  
2                         [jazz, metal]  
3       [reggae, pop, electronic_music]  
4                   [metal, soul, rock]  
...                                 ...  
9995                  [pop, trap, soul]  
9996                     [pop, hip_hop]  
9997  [hip_hop, electronic_music, rock]  
9998                       [trap, rock]  
9999                [funk, disco, soul]  

[10000 rows x 4 columns]

Storing artisits with missing information in a variable

In [ ]:
missing_data_df = pre_warehouse_df[pre_warehouse_df.isnull().any(axis=1)]

Dropping Artists with missing information from pre_warehouse

In [ ]:
pre_warehouse_df = pre_warehouse_df.dropna()

pre_warehouse_df.shape

(9998, 4)

# Data Extraction

Reading client information from txt file

NOTE: This is for security of information

In [ ]:
# Reading credentials.txt file
with open('credentials.txt', 'r') as file:
    lines = file.readlines()

# Extracting client_id and client_secret
client_id = lines[0].strip()
client_secret = lines[1].strip()

print('client information for spotify API successfully loaded ✅')


client information for spotify API successfully loaded ✅


Engineering POST Request to Spotify API token endpoint

In [ ]:
# The endpoint URL
token_url = 'https://accounts.spotify.com/api/token'

# Set up of the data to be sent in the POST request
data = {
    'grant_type': 'client_credentials',
    'client_id': client_id,
    'client_secret': client_secret
}

# Making POST request
response = requests.post(token_url, data=data)

access_token = response.json()['access_token']

print('code ran and access token obtained ✅')


code ran and access token obtained ✅


Example of outlook of response from spotify artist endpoint

In [ ]:
headers = {
    'Authorization': f'"Authorization": "Bearer {access_token}',
}

artist_id = spotify_data_df['spotify_id'][9999]
url = f'https://api.spotify.com/v1/artists/{artist_id}'

response = requests.get(url, headers=headers)

if response.status_code == 200:
    print("Request successful!")
    print(f"Content: {response.content}")
else:
    print(f"Error: {response.status_code}") #error handling



Request successful!
Content: b'{\n  "external_urls" : {\n    "spotify" : "https://open.spotify.com/artist/5wKycvCDB0qE5vvw8CE13M"\n  },\n  "followers" : {\n    "href" : null,\n    "total" : 108\n  },\n  "genres" : [ ],\n  "href" : "https://api.spotify.com/v1/artists/5wKycvCDB0qE5vvw8CE13M",\n  "id" : "5wKycvCDB0qE5vvw8CE13M",\n  "images" : [ {\n    "height" : 640,\n    "url" : "https://i.scdn.co/image/ab6761610000e5eb2d2489e3bfe9f24cd3e556c5",\n    "width" : 640\n  }, {\n    "height" : 320,\n    "url" : "https://i.scdn.co/image/ab676161000051742d2489e3bfe9f24cd3e556c5",\n    "width" : 320\n  }, {\n    "height" : 160,\n    "url" : "https://i.scdn.co/image/ab6761610000f1782d2489e3bfe9f24cd3e556c5",\n    "width" : 160\n  } ],\n  "name" : "Katcross",\n  "popularity" : 0,\n  "type" : "artist",\n  "uri" : "spotify:artist:5wKycvCDB0qE5vvw8CE13M"\n}'


Setting up the Spotify client credentials manager for longer session easy access

In [ ]:
credentials_manager = SpotifyClientCredentials(
    client_id= client_id, client_secret= client_secret
    )

#for connection:
sp = spotipy.Spotify(
    client_credentials_manager=credentials_manager
    )


Extracting all Groover Artsits updated spotify_id through spotify API

In [ ]:
# Let 'artist_name' column be a variable
artist_names = pre_warehouse_df['artist_name']

# Initializing an empty list to store results
spotify_ids = []
popularity_scores = []

# Retrieving Spotify ID and popularity score for each artist
for artist_name in artist_names:
    results = sp.search(q=artist_name, type='artist', limit=1)
    if results['artists']['items']:
        artist_info = results['artists']['items'][0]
        spotify_ids.append(artist_info['id'])
        popularity_scores.append(artist_info['popularity'])
    else:
        # Handling cases where artist name is not found
        spotify_ids.append(None)
        popularity_scores.append(None)

# Adding the retrieved data back to pre_warehouse
pre_warehouse_df['new_spotify_id'] = spotify_ids
pre_warehouse_df['popularity_score'] = popularity_scores

pre_warehouse_df

<ipython-input-20-7fea66cf5029>:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df['new_spotify_id'] = spotify_ids
<ipython-input-20-7fea66cf5029>:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df['popularity_score'] = popularity_scores


user_id     artist_name              spotify_id  \
0       24824     tatum quinn  3s0DwmaExsRr8KGfE8RkhH   
1      303724  jason kerrison  7iLGqGUSoPQtj8OH61HFwZ   
2       28027            k.b.  3Y0w7w2dUKo3DqPh2BDIxr   
3      451943   milo x kahefa  70o5L3YGkE7lxVWPro3fkV   
4      215894          cutter  0ti5jOsC9Jk3TjG5odyaps   
...       ...             ...                     ...   
9995    18629           naeko  0pyq140jTsiXoxlv1KKYzK   
9996    24484  sinatra royale  6avD4d96Qd6SHW70V0fawF   
9997    18645  the rosy cross  0o6kFVOKDFH3MIJUVXkdq1   
9998    18611   didikongstyle  5jnSp9sNhdHJynWddUTdU8   
9999    18618        katcross  5wKycvCDB0qE5vvw8CE13M   

                                  genre          new_spotify_id  \
0                  [reggae, funk, trap]  26PJAv5ZEtZHAi0BzBIXq4   
1                  [reggae, jazz, trap]  3djU8LmyEmiBCNqMCNV688   
2                         [jazz, metal]  007OSCZAlfgm8aX2ZrRyau   
3       [reggae, pop, electronic_music]  05KycaG0g0nYoTk7TjtMSW   
4                   [metal, soul, rock]  22tPgGRQSyj7IdSgZKJyAJ   
...                                 ...                     ...   
9995                  [pop, trap, soul]  0pyq140jTsiXoxlv1KKYzK   
9996                     [pop, hip_hop]  6avD4d96Qd6SHW70V0fawF   
9997  [hip_hop, electronic_music, rock]  4gAWlNtMIGaE4VhzbNq7hc   
9998                       [trap, rock]  5jnSp9sNhdHJynWddUTdU8   
9999                [funk, disco, soul]  5wKycvCDB0qE5vvw8CE13M   

      popularity_score  
0                  0.0  
1                  5.0  
2                 19.0  
3                  0.0  
4                  1.0  
...                ...  
9995              21.0  
9996               0.0  
9997              36.0  
9998               0.0  
9999               0.0  

[9998 rows x 6 columns]

Saving as Parquet file for computational time consevation

In [ ]:
pre_warehouse_df.to_parquet('merged_df.parquet')

In [ ]:
merged_df = pd.read_parquet('merged_df.parquet')

In [ ]:
merged_df

user_id     artist_name              spotify_id  \
0       24824     tatum quinn  3s0DwmaExsRr8KGfE8RkhH   
1      303724  jason kerrison  7iLGqGUSoPQtj8OH61HFwZ   
2       28027            k.b.  3Y0w7w2dUKo3DqPh2BDIxr   
3      451943   milo x kahefa  70o5L3YGkE7lxVWPro3fkV   
4      215894          cutter  0ti5jOsC9Jk3TjG5odyaps   
...       ...             ...                     ...   
9995    18629           naeko  0pyq140jTsiXoxlv1KKYzK   
9996    24484  sinatra royale  6avD4d96Qd6SHW70V0fawF   
9997    18645  the rosy cross  0o6kFVOKDFH3MIJUVXkdq1   
9998    18611   didikongstyle  5jnSp9sNhdHJynWddUTdU8   
9999    18618        katcross  5wKycvCDB0qE5vvw8CE13M   

                                  genre          new_spotify_id  \
0                  [reggae, funk, trap]  26PJAv5ZEtZHAi0BzBIXq4   
1                  [reggae, jazz, trap]  3djU8LmyEmiBCNqMCNV688   
2                         [jazz, metal]  007OSCZAlfgm8aX2ZrRyau   
3       [reggae, pop, electronic_music]  05KycaG0g0nYoTk7TjtMSW   
4                   [metal, soul, rock]  22tPgGRQSyj7IdSgZKJyAJ   
...                                 ...                     ...   
9995                  [pop, trap, soul]  0pyq140jTsiXoxlv1KKYzK   
9996                     [pop, hip_hop]  6avD4d96Qd6SHW70V0fawF   
9997  [hip_hop, electronic_music, rock]  4gAWlNtMIGaE4VhzbNq7hc   
9998                       [trap, rock]  5jnSp9sNhdHJynWddUTdU8   
9999                [funk, disco, soul]  5wKycvCDB0qE5vvw8CE13M   

      popularity_score  
0                  0.0  
1                  5.0  
2                 19.0  
3                  0.0  
4                  1.0  
...                ...  
9995              21.0  
9996               0.0  
9997              36.0  
9998               0.0  
9999               0.0  

[9998 rows x 6 columns]

Checking for count of Groover Artists not on spotify

In [ ]:
merged_df.isnull().sum()

user_id             0
artist_name         0
spotify_id          0
genre               0
new_spotify_id      3
popularity_score    3
dtype: int64

Details of missing Artists missing on spotify

In [ ]:
null_rows = merged_df[merged_df.isnull().any(axis=1)]

null_rows

user_id artist_name              spotify_id                   genre  \
3961   137487       xiz j  6jsBOGztP0Ognxk23GG8Mk             [pop, rock]   
6046     7035       shu28  4eZeECNCDdY9vXkbllYx13  [jazz, hip_hop, disco]   
9078    17202     aeufill  0K1XBsGiqWDwLHQ2ubi0ID     [funk, disco, soul]   

     new_spotify_id  popularity_score  
3961           None               NaN  
6046           None               NaN  
9078           None               NaN

Determining differences in spotify_id between Groover data and spotify data

In [ ]:
# Comparing the two columns to find identical instances
identical = merged_df[merged_df['spotify_id'] == merged_df['new_spotify_id']]

# Counting the number of identical instances
num_identical = identical.shape[0]

# Counting the number of different instances
num_different = merged_df.shape[0] - num_identical

print(f"Number of identical instances of spotify id: {num_identical}")
print(f"Number of different instances of spotify id: {num_different}")


Number of identical instances of spotify id: 4643
Number of different instances of spotify id: 5355


Creating a DataFrame for all data with problems as we cannot just disppose them

In [ ]:
# Artists without ids on spotify
missing_on_spotify = null_rows.iloc[:, :4]

# Concatenating missing data artists with missing on spotify artists
problematic_data = pd.concat(
    [missing_data_df, missing_on_spotify], ignore_index=True)

problematic_data

user_id artist_name              spotify_id                   genre
0     9925         NaN  19mlHo77D3iwhwOA6WB1Hv           [jazz, metal]
1    13997         NaN  42HNwum3iGgcAu583Fsgt4           [reggae, pop]
2   137487       xiz j  6jsBOGztP0Ognxk23GG8Mk             [pop, rock]
3     7035       shu28  4eZeECNCDdY9vXkbllYx13  [jazz, hip_hop, disco]
4    17202     aeufill  0K1XBsGiqWDwLHQ2ubi0ID     [funk, disco, soul]

# <font color = 'gold'> Deliverables Part 1

## <font color= 'gold'>READ ME!
<font color= 'gold'>The DataFrame below, named merged_df, as seen in the Data Extraction section of this notebook, contains the rquirements from the first part of the challenge:

1. <font color= 'gold'>  Each Artist (Groover user) with their popularity score as extracted from spotify

2.  <font color= 'gold'> All (Groover) artist names match with those on spotify, however there are discrepancies between spotify_id information on Groover Database and spotify_id information of the same artists as extracted from spotify.

<font color= 'gold'>We see that 5,355 of those spotify_ids were different.

The approach to go would have been to update Groover database, but this is not within the scope of this work.
</font>


In [ ]:
merged_df

user_id     artist_name              spotify_id  \
0       24824     tatum quinn  3s0DwmaExsRr8KGfE8RkhH   
1      303724  jason kerrison  7iLGqGUSoPQtj8OH61HFwZ   
2       28027            k.b.  3Y0w7w2dUKo3DqPh2BDIxr   
3      451943   milo x kahefa  70o5L3YGkE7lxVWPro3fkV   
4      215894          cutter  0ti5jOsC9Jk3TjG5odyaps   
...       ...             ...                     ...   
9995    18629           naeko  0pyq140jTsiXoxlv1KKYzK   
9996    24484  sinatra royale  6avD4d96Qd6SHW70V0fawF   
9997    18645  the rosy cross  0o6kFVOKDFH3MIJUVXkdq1   
9998    18611   didikongstyle  5jnSp9sNhdHJynWddUTdU8   
9999    18618        katcross  5wKycvCDB0qE5vvw8CE13M   

                                  genre          new_spotify_id  \
0                  [reggae, funk, trap]  26PJAv5ZEtZHAi0BzBIXq4   
1                  [reggae, jazz, trap]  3djU8LmyEmiBCNqMCNV688   
2                         [jazz, metal]  007OSCZAlfgm8aX2ZrRyau   
3       [reggae, pop, electronic_music]  05KycaG0g0nYoTk7TjtMSW   
4                   [metal, soul, rock]  22tPgGRQSyj7IdSgZKJyAJ   
...                                 ...                     ...   
9995                  [pop, trap, soul]  0pyq140jTsiXoxlv1KKYzK   
9996                     [pop, hip_hop]  6avD4d96Qd6SHW70V0fawF   
9997  [hip_hop, electronic_music, rock]  4gAWlNtMIGaE4VhzbNq7hc   
9998                       [trap, rock]  5jnSp9sNhdHJynWddUTdU8   
9999                [funk, disco, soul]  5wKycvCDB0qE5vvw8CE13M   

      popularity_score  
0                  0.0  
1                  5.0  
2                 19.0  
3                  0.0  
4                  1.0  
...                ...  
9995              21.0  
9996               0.0  
9997              36.0  
9998               0.0  
9999               0.0  

[9998 rows x 6 columns]

# <font color= 'gold'> Deliverables Part 2 (Additional Question)

Installing the postgreSQL interface dependencies

In [ ]:
%pip install psycopg2 ipython-sql

Note: you may need to restart the kernel to use updated packages.


ERROR: Could not find a version that satisfies the requirement ipython-sql (from versions: none)
ERROR: No matching distribution found for ipython-sql


In [ ]:
%load_ext sql

Creating a connection url

In [ ]:
# Extracting password from credentials.txt
pword = lines[2].strip()


In [ ]:
connection = f'postgresql://postgres:{pword}@localhost:5432/Groover_db'

In [ ]:
%sql $connection

SQL query to fetch: user_id, spotify_id, genres by artist, total number of genres artist is listed in and total number of artists listed in each genre (displaying first 25 rows)

In [ ]:
%%sql
-- Step 1: Aggregate genres for each user
WITH aggregated_genres AS (
    SELECT
        tad.user_id,
        STRING_AGG(tgd.genre, ', ') AS genres_by_artist
    FROM
        tag_artist_data tad
    JOIN
        tag_genre_data tgd ON tad.tag_id = tgd.tag_id
    GROUP BY
        tad.user_id
),

-- Step 2: Count genres per user
count_genres AS (
    SELECT
        tad.user_id,
        COUNT(tad.tag_id) AS genres_per_user
    FROM
        tag_artist_data tad
    GROUP BY
        tad.user_id
),

-- Step 3: Combine user data with aggregated genres and genre counts
artists_info_query AS (
    SELECT
        sd.user_id,
        sd.spotify_id,
        ag.genres_by_artist,
        cg.genres_per_user,
        ROW_NUMBER() OVER () AS row_num
    FROM
        spotify_data sd
    LEFT JOIN
        aggregated_genres ag ON sd.user_id = ag.user_id
    LEFT JOIN
        count_genres cg ON sd.user_id = cg.user_id
    LIMIT 25
),

-- Step 4: Aggregate artist counts by genre
genre_info_query AS (
    SELECT
        tgd.genre,
        COUNT(DISTINCT tad.user_id) AS artist_in_genre,
        ROW_NUMBER() OVER () AS row_num
    FROM
        tag_artist_data tad
    JOIN
        tag_genre_data tgd ON tad.tag_id = tgd.tag_id
    GROUP BY
        tgd.genre
    LIMIT 11
)

-- Step 5: Combine the two queries using their row numbers
SELECT
    aiq.user_id,
    aiq.spotify_id,
    aiq.genres_by_artist,
    aiq.genres_per_user,
    giq.genre,
    giq.artist_in_genre
FROM
    artists_info_query aiq
LEFT JOIN
    genre_info_query giq ON aiq.row_num = giq.row_num;


 * postgresql://postgres:***@localhost:5432/Groover_db
25 rows affected.


[(9, '5e2WCQCvRUo05S2uTk2xVC', 'jazz, electronic_music, reggae', 3, 'disco', 2473),
 (12, '7MOpb0hgwnTxr3lCNPPVGR', 'soul, rock', 2, 'electronic_music', 2456),
 (14, '6VcXjtZBueCzpqWFqg29O7', 'pop, trap, funk', 3, 'funk', 2594),
 (16, '2hYPsr25gOfRQCsz7Boe1Q', 'electronic_music, soul', 2, 'hip_hop', 2476),
 (23, '1pXMpZ5naNbGArl4Q1DGhs', 'trap, disco, rock', 3, 'jazz', 2442),
 (27, '1M7FGvgTp0ftLNZiKS61fp', 'jazz, electronic_music, funk', 3, 'metal', 2523),
 (31, '3GznvHxIVOUCw9pmG76Nno', 'hip_hop, jazz, trap', 3, 'pop', 2448),
 (33, '2vjzdOU9wy0flWHeQI3d6c', 'funk, soul, rock', 3, 'reggae', 2448),
 (34, '4vBjqyZSwW1mee1gXfo9Cv', 'jazz, soul, rock', 3, 'rock', 2401),
 (40, '4TJVX2FfoQKIVESztwvwVS', 'metal, funk, rock', 3, 'soul', 2540),
 (44, '2YEFQTXVGZSspfPFeFiDrl', 'reggae, trap, soul', 3, 'trap', 2500),
 (61, '4Y7lEpGZL5AsgZhzjBjYya', 'electronic_music, metal, rock', 3, None, None),
 (65, '0ssl0AzN5WYHeyBvzCNy4d', 'jazz, pop', 2, None, None),
 (67, '3qKl1S1KWf6j14Jg31M1Li', 'jazz, metal, trap', 3, None, None),
 (72, '5PJrA4AsPWMQVKvHYcFqsp', 'hip_hop, metal, funk', 3, None, None),
 (85, '4aqFxUUZMhGXJhyKIMmHQU', 'hip_hop, metal, funk', 3, None, None),
 (88, '4OCjOBEzP8yTUPh6puX8ly', 'electronic_music, trap, disco', 3, None, None),
 (94, '2attj72mDzqGW51nnkdpuc', 'hip_hop, pop, soul', 3, None, None),
 (95, '09XlGeIi4ig81GQDaec7i6', 'trap, funk, rock', 3, None, None),
 (101, '4RvaqCAeKu0wKQ8Sliz3AZ', 'jazz, disco', 2, None, None),
 (103, '33aFYq4gUZ8v8XyXeOjt69', 'trap, funk', 2, None, None),
 (106, '113JeJYtNW38BCbXzNLXKq', 'jazz, trap', 2, None, None),
 (111, '6hmYmasgOBfCcSAsZApRr1', 'electronic_music, metal, disco', 3, None, None),
 (113, '486eivvjMrA8ZyjYzlie10', 'jazz, electronic_music, reggae', 3, None, None),
 (119, '7F1N7XfIeMubysW5dUmekt', 'metal, funk, rock', 3, None, None)]

# <font color = 'blue'>THANKS FOR YOUR LISTENING!